---
title: "Bioconductor-IRanges-n-GenomicRanges"
author: "Hira Ali"
date: "2025-02-15"
output: html_document
---

ExpressionSet was designed for microarray-based experiments, in which genomic features to be measured are all prespecified and can be individually named in advance.  

With NGS, features of interest can be defined fairly arbitrarily in terms of genomic coordinates.  Additional constraints are important, so that the reference within which coordinates are defined is made explicit, and the compatibility of region specification with reference limitations (for example, upper bounds on addresses given by chromosome lengths) can be guaranteed.  

In this Rmd, I'll overview Bioconductor's approach to managing sequencing-based assays depends fundamentally on the IRanges and GenomicRanges packages.

# Interval ranges: IRanges

## Install & Load IRanges package

```{r}
BiocManager::install("IRanges")
library(IRanges)
```

## Basics

Demonstrating some of the most key functions:

```{r}

ir <- IRanges(5,10)
ir

```

```{r}
start(ir)
end(ir)
width(ir)
```

A single IRanges object can hold more than one range. We do this by specifying vector to the start and end arguments.

```{r}
# specifying the starting value of the vectors, the ending value of the vectors, and will also output the width from start to end.
IRanges(start=c(3,5,17), end=c(10,8,20))

```

For example in the first row, we see a start at "base pair" 3, ending at "base pair" 10, and the width is 8 "base pairs long".

```{r}
ir
ir <- IRanges(5, 10)
```

There are also various intra-range methods. Intra-range means that the operation will occur for each range that you have and it doesn't depend on the other ranges that are contained in the IRanges object.

```{r}

# i.e. we can shift the IRange by 2
shift(ir, -2)

```
Now, we have an output of a range that starts at 3 (instead of 5 originally) and ends at 8 (originally ended at 10).

```{r}

# i.e. narrow() says relative to the start, you should start this range at the 2nd base pair
narrow(ir, start=2)

# i.e. relative to the end, you should end at the 5th base pair
narrow(ir, end = 5)
```
```{r}

# flank() allows you to get flanking sequence here, 3 base pairs from the start 
flank(ir, width=3, start=TRUE, both=FALSE)

# you can also get 3 base pairs from the end by specifying start = FALSE
flank(ir, width=3, start=FALSE, both=FALSE)


```
## Seeing the same basic functions Graphically

Those same operations plotted in a single window. The red bar shows the shadow of the original range ir.

```{r}
# set up a plotting window so we can look at range operations
plot(0,0,xlim=c(0,23),ylim=c(0,13),type="n",xlab="",ylab="",xaxt="n")
axis(1,0:15)
abline(v=0:14 + .5,col=rgb(0,0,0,.5))

# plot the original IRange
## number indicates the base pair
plotir <- function(ir,i) { arrows(start(ir)-.5,i,end(ir)+.5,i,code=3,angle=90,lwd=3) }
plotir(ir,1)

# draw a red shadow for the original IRange
polygon(c(start(ir)-.5,start(ir)-.5,end(ir)+.5,end(ir)+.5),c(-1,15,15,-1),col=rgb(1,0,0,.2),border=NA)

# draw the different ranges
plotir(shift(ir,-2), 2)
plotir(narrow(ir, start=2), 3)
plotir(narrow(ir, end=5), 4)
plotir(flank(ir, width=3, start=TRUE, both=FALSE), 5)
plotir(flank(ir, width=3, start=FALSE, both=FALSE), 6)
plotir(flank(ir, width=3, start=TRUE, both=TRUE), 7)
plotir(ir * 2, 8)
plotir(ir * -2, 9)
plotir(ir + 2, 10)
plotir(ir - 2, 11)
plotir(resize(ir, 1), 12)

text(rep(15,12), 1:12, c("ir","shift(ir,-2)","narrow(ir,start=2)",
                         "narrow(ir,end=5)",
                         "flank(ir, start=T, both=F)",
                         "flank(ir, start=F, both=F)",
                         "flank(ir, start=T, both=T)",
                         "ir * 2","ir * -2","ir + 2","ir - 2",
                         "resize(ir, 1)"), pos=4)

```

## Example: Creating an IRanges object with 3 ranges

```{r}
# full details on the inter-range methods:
# ?"inter-range-methods"
(ir <- IRanges(start=c(3,5,17), end=c(10,8,20)))
range(ir)
```

The `reduce` function collapses the ranges, so that integers are covered by only one range in the output.

```{r}
reduce(ir)
```

The `gaps` function gives back the ranges of integers which are in `range(ir)` but not covered by any of the ranges in `ir`:

```{r}
gaps(ir)
```

The `disjoin` function breaks up the ranges in `ir` into discrete ranges. This is best explained with examples, but here is the formal definition first:

> returns a disjoint object, by finding the union of the
> end points in ‘x’. In other words, the result consists of a range
> for every interval, of maximal length, over which the set of
> overlapping ranges in ‘x’ is the same and at least of size 1.

```{r}
disjoin(ir)
```
# Genomic Ranges: GRanges

Genomic Ranges is an extension of IRanges to the Genomic Space.

## Install & Load GRanges Package

```{r}
BiocManager::install("GenomicRanges")
library(GenomicRanges)
```

*GRanges* are objects which contain *IRanges* and two more important pieces of information:

* the chromosome we are referring to (called `seqnames` in Bioconductor)
* the strand of DNA we are referring to

Strand can be specified as plus "+" or minus "-", or left unspecified with "\*". 

* Plus strand features have the biological direction from left to right on the number line, * Minus strand features have the biological direction from right to left.

In terms of the *IRanges*:

* plus strand features go from `start` to `end`
* minus strand features go from `end` to `start`. 

This is required because `width` is defined as `end - start + 1`, and negative width ranges are not allowed. Because DNA has two strands, which have an opposite directionality, strand is necessary for uniquely referring to DNA.

With an *IRange*, a chromosome name, and a strand, we can be sure we are uniquely referring to the same region and strand of the DNA molecule as another researcher, given that we are using the same build of *genome*. There are other pieces of information which can be contained within a GRanges object, but the two above are the most important.

## Example: Practice with two ranges with a made-up chromosome

Creationg a set of two ranges on a made-up chromosome, *chrZ*. These ranges refer to the genome *hg19*. Because we have not linked our genome to a database, we are allowed to specify a chromosome which does not really exist in *hg19*.

```{r}
# using GRanges to create an object called gr. naming the chromosome chrZ.
# using IRanges to define the start and end positions of the selected regions on the chromosome. AKA range of a set of sequences.
## in this case, we have two sets, first from position 5 to 35, and the second from position 10 to 45.
gr <- GRanges("chrZ", IRanges(start=c(5,10),end=c(35,45)),
              # specifying strand feature: + direction
              # specifying the total length of chromosome Z = 100 base pairs.
              strand="+", seqlengths=c(chrZ=100L))
# when returning this initially, we see that seqinfo mentions that the genome is unspecified.
gr

# specifying the genome of object, gr (from Human Genome reference)
## Recall that because awe haven't actually linked our genome to a datbase, we can specify a chromosome that doesn't actually exist in hg19.
genome(gr) <- "hg19"
# now returns seqinfo: 1 sequence from hg19 genome
gr
```
Note the `seqnames` and `seqlengths` which we defined in the call above:

```{r}
seqnames(gr)
seqlengths(gr)
```
We can use the `shift` function as we did with the IRanges. However, notice the warning when we try to shift the range beyond the length of the chromosome:

```{r}
# shifting ranges up 10 (i.e. from start = 5 to start = 15)
shift(gr, 10)
# this wont work bc it would be out of range for the chromosome length (i.e. from end = 35 to end = 115 which doesnt exist)
shift(gr, 80)
```

If we `trim` the ranges, we obtain the ranges which are left, disregarding the portion that stretched beyond the length of the chromosome:

```{r}
# using trim, we wont go out of bounds for the range if the position does not exist
trim(shift(gr, 80))
```

We can add columns of information to each range using the `mcols` function (stands for *metadata columns*). Note: this is also possible with IRanges. We can remove the columns by assigning `NULL`.

```{r}
# at this point, will return 2 rows and 0 columns since we haven't added any metadata
mcols(gr)

# at this point, we have added 1 metadata column named 'value' and specified the values in the new column
mcols(gr)$value <- c(-1,4)
gr

# specifying 'NULL' removes the values associated in the created columnm
mcols(gr)$value <- NULL
```

### GRangesList

Especially when referring to genes, it is useful to create a *list* of GRanges. This is useful for representing groupings, for example the [exons](http://en.wikipedia.org/wiki/Exon) which belong to each gene. The elements of the list are the genes, and within each element the exon ranges are defined as GRanges.

```{r}
# Creating a second GRanges object called gr2
## "chrZ" specifies that the following ranges belong to chromosome Z.
## 11:13 is specifying all start positions between 11:13 (inclusive), AKA start positions are 11, 12, and 13. Same for 51:53 (specifying the end positions).
## Note: IRanges() pairs element-wise
gr2 <- GRanges("chrZ",IRanges(11:13,51:53))

# now we can create a new GRanges list by wrapping these two objects using the GRangesList function
grl <- GRangesList(gr, gr2)
# Now this object contains two GRanges.
grl
```

The most obvious example of a GRangesList would be grouping exons by gene, or grouping exons by transcript.

The length of the *GRangesList* is the number of *GRanges* object within. To get the length of each GRanges we call `elementNROWS`. We can index into the list using typical list indexing of two square brackets.

```{r}
# Returns the number of elements (AKA the number of GRanges) in grl. In this case, we have 2 GRanges objects.
length(grl)

# Returns the number of genomic ranges within each element (i.e., the number of rows in each GRanges object). In this case, the first GRanges object has 2 ranges. The second GRanges elemenet has 3 ranges.
elementNROWS(grl)

# To return the first element, which would be the first GRanges object.
grl[[1]]

```
If we ask the `width`, the result is an *IntegerList*. If we apply `sum`, we get a numeric vector of the sum of the widths of each GRanges object in the list.

```{r}
# telling us the length of each of the ranges, within each of the GRanges objects.
## i.e. first element of first GRanges object is of start = 5 & end = 35. so the  width is = end - start + 1 = 35 - 5 + 1 = 31. This is the first value returned of the first GRanges object.
width(grl)
# adding the lengths of ranges of each GRanges object together (i.e. for first object, 31 + 36 = 67, which is the first value returned).
sum(width(grl))
```

We can add metadata columns as before, now one row of metadata for each GRanges object, not for each range. It doesn't show up when we print the GRangesList, but it is still stored and accessible with `mcols`.

```{r}
mcols(grl)$value <- c(5,7)
grl
mcols(grl)
```

### findOverlaps and %over%

We will demonstrate two commonly used methods for comparing GRanges objects. First we build two sets of ranges:

```{r}
(gr1 <- GRanges("chrZ",IRanges(c(1,11,21,31,41),width=5),strand="*"))
(gr2 <- GRanges("chrZ",IRanges(c(19,33),c(38,35)),strand="*"))
```

`findOverlaps` returns a *Hits* object which contains the information about which ranges in the query (the first argument) overlapped which ranges in the subject (the second argument). There are many options for specifying what kind of overlaps should be counted.

```{r}
# creating an object called fo (find objects). query = gr1 (first argument), and subject = gr2 (second argument)
fo <- findOverlaps(gr1, gr2)
# i.e. in this case, we see a total of 3 hits between the query and subject.
## i.e. in this case, we see that range 3 (AKA 3rd element) of gr1 (21-25) overlaps with range 1 of gr2 (19-38)
fo
# returns the indices of the query regions that have an overlap (i.e. if an entry in query overlaps multiple entries in subject, it will appear multiple times)
queryHits(fo)
# returns the indicies of subject regions that overlap with the corresponding queryHits(). i.e. if a query overlaps multiple subject ranges, you get multiple indicies.
subjectHits(fo)
```
Another way of getting at overlap information is to use `%over%` which returns a logical vector of which ranges in the first argument overlapped any ranges in the second.

```{r}
# returns TRUE or FALSE, where TRUE is stating that the range in the first argument overlapped with a range in the second.
gr1 %over% gr2
# returning only ranges in gr1 that had some overlap with gr2
gr1[gr1 %over% gr2]
```
Note that both of these are *strand-specific*, although `findOverlaps` has an `ignore.strand` option.

```{r}
gr1 <- GRanges("chrZ",IRanges(1,10),strand="+")
gr2 <- GRanges("chrZ",IRanges(1,10),strand="-")
gr1 %over% gr2
```

### Rle and Views

Lastly, here is a short glimpse into two related classes defined in IRanges, the *Rle* and *Views* classes. 

* *Rle* stands for *run-length encoding*, which is a form of compression for repetitive data. Instead of storing: $[1,1,1,1]$, we would store the number 1, and the number of repeats: 4. The more repetitive the data, the greater the compression with *Rle*.

Some notes: 
* Runs = unique values
* Lengths = How many times each value repeats

We use `str` to examine the internal structure of the Rle, to show it is only storing the numeric values and the number of repeats

```{r}
# Rle() create a Run-Length Encoding object, which is a way to store repeated values efficiently. Representing values in a compact form.
## The argument inside Rle() is a vector containing repeated values.
## rep() expands a value x, y # of times. i.e. expanding the value, -1, twenty times.
## Note, return includes the total length which = listed values + 20 repeats = 28. It also includes the # of runs which is 4 (values include 1, 0, -2, and -1).
## () around the whole line prints r immediately.
(r <- Rle(c(1,1,1,0,0,-2,-2,-2,rep(-1,20))))

# shows the internal structure of the Rle object. This includes the unique values, the lengths of each of the unique values respectively, the elementMetadata, and the metadata.
## Only storing the compact representation.
str(r)

# as.numeric() converts Rle object, r, back into a regular numeric vector. 
## Now, we would be able to return the expanded version.
as.numeric(r)
```

A *Views* object can be thought of as "windows" looking into a sequence. Or in other words, if we want to view different regions of the sequence.

* *Views* is a virtual class which contains the subject, and then the number of views which are essentially IRanges into that object. It is a virtual class in the sense that it does not deconstruct the Rle, but instead we have just saved the Rle and a number of windows into it.

```{r}
# i.e. we have the Rle object, r. We want to create two views, one which starts at 4 and ends at 7, and another that starts at 2 and ends at 6.
# returns the values at these positions within the Rle.
(v <- Views(r, start=c(4,2), end=c(7,6)))
```

Note that the internal structure of the Views object is just the original object, and the *IRanges* which specify the windows. The great benefit of Views is when the original object is not stored in memory, in which case the Views object is a lightweight class which helps us reference subsequences, without having to load the entire sequence into memory.

 You can also use the Views constructor for FASTA files, for example, if you want to look into genome sequence or other objects.

```{r}
# Looking at the internal structure of the object, v, of virtual class or in other words, formal class is of Rle.
str(v)
```